File: Warehouse.Backend/Core/Infrastructure/MarketDataCache.cs
using System.Collections.Concurrent;
using System.Collections.Frozen;
using Warehouse.Backend.Core.Abstractions.Markets;
using Warehouse.Backend.Core.Domain;
using Warehouse.Backend.Core.Models;

namespace Warehouse.Backend.Core.Infrastructure;

public class MarketDataCache(ILogger<MarketDataCache> logger) : IMarketDataCache
{
    private readonly ConcurrentDictionary<MarketDataKey, MarketDataSnapshot> cache = new();

    public MarketData? GetData(string symbol, MarketType marketType)
    {
        if (cache.TryGetValue(new MarketDataKey(symbol, marketType), out MarketDataSnapshot? snapshot))
        {
            return snapshot.ToMarketData();
        }

        return null;
    }

    public void Update(MarketDataEvent marketDataEvent)
    {
        MarketDataSnapshot snapshot = cache.GetOrAdd(
            new MarketDataKey(marketDataEvent.Symbol, marketDataEvent.Source),
            _ => new MarketDataSnapshot());

        foreach (string[] ask in marketDataEvent.Asks)
        {
            if (ask.Length >= 4 &&
                decimal.TryParse(ask[0], out decimal price) &&
                decimal.TryParse(ask[1], out decimal size) &&
                int.TryParse(ask[3], out int orderCount))
            {
                if (size == 0)
                {
                    snapshot.Asks.TryRemove(price, out _);
                }
                else
                {
                    snapshot.Asks[price] = (size, orderCount);
                }
            }
        }

        foreach (string[] bid in marketDataEvent.Bids)
        {
            if (bid.Length >= 4 &&
                decimal.TryParse(bid[0], out decimal price) &&
                decimal.TryParse(bid[1], out decimal size) &&
                int.TryParse(bid[3], out int orderCount))
            {
                if (size == 0)
                {
                    snapshot.Bids.TryRemove(price, out _);
                }
                else
                {
                    snapshot.Bids[price] = (size, orderCount);
                }
            }
        }
    }

    private record MarketDataKey(string Symbol, MarketType MarketType);

    private class MarketDataSnapshot
    {
        public ConcurrentDictionary<decimal, (decimal Size, int OrderCount)> Asks { get; } = new();

        public ConcurrentDictionary<decimal, (decimal Size, int OrderCount)> Bids { get; } = new();

        public MarketData ToMarketData() => new(Asks.ToFrozenDictionary(), Bids.ToFrozenDictionary());
    }
}

File: src/Warehouse.Core/Markets/MarketDataCache.fs
namespace Warehouse.Core.Markets.Services

open System
open System.Collections.Concurrent
open System.Collections.Frozen
open System.Globalization
open Warehouse.Core.Markets.Contracts
open Warehouse.Core.Markets.Domain

[<Struct>]
type private MarketDataKey = { Symbol: string; MarketType: MarketType }

type private MarketDataSnapshot() =
    member val Asks = ConcurrentDictionary<decimal, struct (decimal * int)>()
    member val Bids = ConcurrentDictionary<decimal, struct (decimal * int)>()

    member this.ToMarketData() = { Asks = this.Asks.ToFrozenDictionary(); Bids = this.Bids.ToFrozenDictionary() }

type MarketDataCache() =
    let cache = ConcurrentDictionary<MarketDataKey, MarketDataSnapshot>()

    let tryParseLevel (data: string[]) =
        if data.Length < 4 then
            ValueNone
        else
            match
                Decimal.TryParse(data[0], CultureInfo.InvariantCulture),
                Decimal.TryParse(data[1], CultureInfo.InvariantCulture),
                Int32.TryParse(data[3], CultureInfo.InvariantCulture)
            with
            | (true, price), (true, size), (true, orderCount) -> ValueSome struct (price, size, orderCount)
            | _ -> ValueNone

    let updateSide (side: ConcurrentDictionary<decimal, struct (decimal * int)>) (levels: string[][]) =
        for level in levels do
            match tryParseLevel level with
            | ValueSome struct (price, size, orderCount) ->
                if size = 0m then side.TryRemove(price) |> ignore else side[price] <- struct (size, orderCount)
            | ValueNone -> ()

    interface IMarketDataCache with
        member _.GetData(symbol, marketType) =
            let key = { Symbol = symbol; MarketType = marketType }

            match cache.TryGetValue(key) with
            | true, snapshot -> Some(snapshot.ToMarketData())
            | false, _ -> None

        member _.Update(marketDataEvent) =
            let key = { Symbol = marketDataEvent.Symbol; MarketType = marketDataEvent.Source }
            let snapshot = cache.GetOrAdd(key, fun _ -> MarketDataSnapshot())

            updateSide snapshot.Asks marketDataEvent.Asks
            updateSide snapshot.Bids marketDataEvent.Bids
